### PR TITLE
User/getrou/converged nuget

### PR DESCRIPTION
**Pull request by getrou**
The new artifact and nuget layout can be seen here:
https://dev.azure.com/ms/ProjectReunion/_build/results?buildId=114478&view=artifacts&type=publishedArtifacts

### CODE REVIEW

File: dev/MRTCore/azure-pipelines.yml [hunk 1]
```diff
@@ -383,29 +383,15 @@ jobs:
...
         mrm.h
-      TargetFolder: '$(Build.ArtifactStagingDirectory)\mrt\include'
+      TargetFolder: '$(Build.ArtifactStagingDirectory)\drop\include'
 
   - task: CopyFiles@2
     displayName: 'copy license'
     inputs:
       SourceFolder: '$(MRTSourcesDirectory)'
       Contents: |
         LICENSE.txt
```
Comment (huichen123) [inline dev/MRTCore/azure-pipelines.yml:393]: >LICENSE.txt [](start = 8, length = 11)

we need to deprecate this one with the one under root folder.

File: dev/MRTCore/azure-pipelines.yml [hunk 2]
```diff
@@ -383,29 +383,15 @@ jobs:
...
-      versioningScheme: 'byPrereleaseNumber'
-      majorVersion: '$(Major)'
-      minorVersion: '$(Minor)'
-      patchVersion: '$(Patch)'
-      basePath: '$(Build.ArtifactStagingDirectory)\mrt'
-      packDestination: '$(Build.ArtifactStagingDirectory)\nuget'
+      TargetFolder: '$(Build.ArtifactStagingDirectory)\drop'
 
   - ${{if eq(parameters.doCodeSign, 'true')}}:
     - task: PkgESCodeSign@10
```
Comment (huichen123) [inline dev/MRTCore/azure-pipelines.yml:397]: this task is no longer needed

File: dev/MRTCore/azure-pipelines.yml [hunk 3]
```diff
@@ -414,12 +400,12 @@ jobs:
         SYSTEM_ACCESSTOKEN: $(System.AccessToken)
       inputs:
         signConfigXml: '$(MRTSourcesDirectory)\SignConfig-nuget.xml'
-        inPathRoot: '$(BUILD.ArtifactStagingDirectory)\nuget'
-        outPathRoot: '$(BUILD.ArtifactStagingDirectory)\nuget'
+        inPathRoot: '$(BUILD.ArtifactStagingDirectory)\drop'
+        outPathRoot: '$(BUILD.ArtifactStagingDirectory)\drop'
   
   - task: PublishBuildArtifacts@1
```
Comment (huichen123) [inline dev/MRTCore/azure-pipelines.yml:406]: >- task: PublishBuildArtifacts@1 [](start = 2, length = 31)

no longer needed

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (huichen123) [inline dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: this file should not be under MRTCore folder

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+<package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
+     <metadata minClientVersion="3.3">
+         <id>Microsoft.ProjectReunion</id>
+         <version>1.0</version>
+         <title>$id$</title>
+         <authors>Microsoft</authors>
+         <description>Resource loader for .PRI files.</description>
```
Comment (huichen123) [inline dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:8]: >Resource loader for .PRI files. [](start = 22, length = 31)

description is not correct now.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+         <license type="file">LICENSE.txt</license>
+         <requireLicenseAcceptance>true</requireLicenseAcceptance>
+         <projectUrl>https://aka.ms/MRTCore</projectUrl>
```
Comment (huichen123) [inline dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:11]: >https://aka.ms/MRTCore [](start = 21, length = 22)

also need to point to project reunion

File: build/ProjectReunion-CI.yml [hunk 7]
```diff
@@ -50,8 +55,9 @@ jobs:
 - template: AzurePipelinesTemplates\ProjectReunion-CreateNugetPackage-Job.yml
   parameters:
     jobName: CreateNugetPackage
-    dependsOn: Build
+    dependsOn: PublishMRT
     prereleaseVersionTag: ci
+    condition: Always()
```
Comment (huichen123) [inline build/ProjectReunion-CI.yml:60]: >condition: Always() [](start = 4, length = 19)

shouldn't do only when succeeded?

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml [hunk 8]
```diff
@@ -53,36 +58,52 @@ jobs:
...
-        -DateOverride '$(builddate)' `
-        -Subversion '$(buildrevision)' `
-        -BuildArch ${{ parameters.primaryBuildArch }} `
-        -BuildFlavor ${{ parameters.buildFlavor }}
-
-    displayName: 'build-nupkg.ps1'
+  - task: 333b11bd-d341-40d9-afcf-b32d5ce6f23b@2
+    inputs:
+      command: 'pack'
+      packagesToPack: '$(Build.SourcesDirectory)\dev\MRTCore\mrt\NuGet\Microsoft.ProjectReunion.nuspec'
```
Comment (huichen123) [inline build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:96]: >MRTCore\mrt\NuGet\Microsoft.ProjectReunion.nuspec [](start = 53, length = 49)

this needs to be outside MRTCore

**Comment by huichen123** [comment]
>         outPathRoot: '${{ parameters.nupkgdir }}\signed'

remove

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:116 in aa04ba8. [](commit_id = aa04ba83695ef955045b453845fe7eaf744c5c8f, deletion_comment = False)

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml [hunk 9]
```diff
@@ -53,36 +58,52 @@ jobs:
...
+      versioningScheme: 'byPrereleaseNumber'
+      majorVersion: ${{ parameters.major }}
+      minorVersion: ${{ parameters.minor }}
+      patchVersion: ${{ parameters.patch }}
+      basePath: '$(Build.SourcesDirectory)\Artifacts\drop'
```
Comment (huichen123) [inline build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:101]: >drop [](start = 53, length = 4)

as is we pack way too much stuff. (ok to fix it later)

**Comment by huichen123** [comment]
@EHO-makai , it appears the pipeline is not run. neither for your PR. something is wrong here.

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml [hunk 10]
```diff
@@ -53,36 +58,52 @@ jobs:
...
-    - task: CopyFiles@2
-      inputs:
-        targetFolder: '${{ parameters.nupkgdir }}'
-        sourceFolder: '$(Build.SourcesDirectory)\Artifacts\drop'
-        Contents: '**\Microsoft.ProjectReunion*.appx'
-
-  - powershell: |
-      $prereleaseTag = "${{ parameters.prereleaseVersionTag }}"
-      if ("${{ parameters.useReleaseTag}}" -eq [bool]::TrueString) { $prereleaseTag = "" }
+  # UNDONE to move MRTCore into Project Reunion nuget package
```
Comment (EHO-makai) [inline build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:66]: are you going to re-enable this before completing this PR?

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml (hunk 8 above)
Comment (EHO-makai) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:96]: Feels a little weird to have 'Microsoft.ProjectReunion.nuspec' in dev\MRTCore.  I suppose this is just a "phase 1" kind of hting.

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml (hunk 9 above)
Comment (EHO-makai) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:101]: we're supposed to pack the exact same thing we were packing before, this is something I told Alan Wu we would do.

Comment (EHO-makai) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:101]: OTOH, I love that we're using the nuget pack command.  :-)

Comment (huichen123) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:101]: if you open the package, you know what i was saying. :)

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml [hunk 11]
```diff
@@ -98,7 +119,7 @@ jobs:
     displayName: 'Publish artifact: nupkg'
     inputs:
       PathtoPublish: '${{ parameters.nupkgdir }}'
-      artifactName: 'drop'
+      artifactName: 'nuget'
```
Comment (EHO-makai) [inline build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:122]: I have a precise directory that I use to do the nuget signing, this will have to be adjusted/re-tested after you completee this PR.

File: build/ProjectReunion-CI.yml (hunk 7 above)
Comment (EHO-makai) [reply build/ProjectReunion-CI.yml:60]: Geoffrey, was this the weird "wouldn't run" problem we were looking at before?  Does seem like we should remove always as a condition.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 4 above)
Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: I think this is more or less true I wonder when we will have a plan for aggregating multiple nuspecs for the sub modules inside of the Project Reunion repo into one we can feed to the nuget pack command.  I think we don't have this plan yet so it's probably OK that it is here temporarily.

Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: You two can work out what name this should have.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 6 above)
Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:11]: I'll have somebody register aka.ms/projectreunion to point to github/microsoft/projectreunion.

File: build/ProjectReunion-CI.yml (hunk 7 above)
Comment (getrou) [reply build/ProjectReunion-CI.yml:60]: Ah, yes.  I still haven't figured out why when this condition runs, succeeded() and succeededOrFailed() both return false.

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml (hunk 10 above)
Comment (getrou) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:66]: No, this removes the creation of the nuget and framework packages for the Project Reunion sample app.  I'm leaving it in as a comment so we have the framework package structure in case we want to use it to build the framework packages for our new MRTCore/ProjectReunion nuget package.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 4 above)
Comment (getrou) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: I also asked this question, and the answer is that MRTCore doesn't have an identity outside of Project Reunion, so it shouldn't get it's own nuget package.  The assertion is that there won't be anyone using MRTCore without using all of ProjectReunion (aside from WinUI who will do something special).

Comment (huichen123) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: when Reunion grows, say it contains 100 subprojects. It's very unlikely someone will need them all.

File: dev/MRTCore/azure-pipelines.yml (hunk 3 above)
Comment (getrou) [reply dev/MRTCore/azure-pipelines.yml:406]: If we don't do this, won't we not have access to the signed binaries in the next job?

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 4 above)
Comment (getrou) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec:1]: @EHO-makai would you like to take this?

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 5 above)
Comment (getrou) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: Fixed, moved file out of this directory and changed description

File: build/NuSpecs/Microsoft.ProjectReunion.nuspec [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
+     <metadata minClientVersion="3.3">
+         <id>Microsoft.ProjectReunion</id>
+         <version>1.0</version>
+         <title>$id$</title>
+         <authors>Microsoft</authors>
+         <description>Project Reunion</description>
```
Comment (huichen123) [inline build/NuSpecs/Microsoft.ProjectReunion.nuspec:8]: >Project Reunion [](start = 22, length = 15)

doesn't sound very helpful :). we can change later though

File: dev/MRTCore/azure-pipelines.yml (hunk 3 above)
Comment (huichen123) [reply dev/MRTCore/azure-pipelines.yml:388]: isn't nuget handled by CreateNugetPackage now?

File: build/NuSpecs/Microsoft.ProjectReunion.nuspec (hunk 12 above)
Comment (getrou) [reply build/NuSpecs/Microsoft.ProjectReunion.nuspec:8]: Sure.  I just copied what it said in the original ProjectReunion nuspec file.  I'm sure a PM will come up with something :)

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml (hunk 9 above)
Comment (huichen123) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:108]: yes.

File: dev/MRTCore/azure-pipelines.yml (hunk 3 above)
Comment (getrou) [reply dev/MRTCore/azure-pipelines.yml:388]: Maybe I misunderstand the point of publishing a build artifact.  The point of this is to push the signed binaries to a "common drop point" that other sub-modules (once they exist) will also publish to.  Then in the CreateNugetPackage job, we download the contents of this drop point and wrap them into a nuget package.

So this step is just pushing the binaries to the common drop point.

Comment (huichen123) [reply dev/MRTCore/azure-pipelines.yml:388]: create/sign/publish nuget are all handled in CreateNugetPackage now.

Comment (getrou) [reply dev/MRTCore/azure-pipelines.yml:388]: Sure, that makes sense.  But this "publish" step is just publishing the signed binaries to an intermediate location so they are available to the next job.
This step isn't publishing the final nuget package, it's just publishing some intermediate binaries so that the next job, which is CreateNugetPackage, can use them.

Comment (huichen123) [reply dev/MRTCore/azure-pipelines.yml:388]: sorry, you are right. they are different jobs, thus need publish in between.

File: build/ProjectReunion-CI.yml (hunk 7 above)
Comment (getrou) [reply build/ProjectReunion-CI.yml:60]: https://dev.azure.com/ms/ProjectReunion/_build/results?buildId=114742&view=results

Here's what happens when I change it to only run on succeeded().  The same thing happens if I use succeededOrFailed().  I'm not sure why, but Always() is the only condition that is working.

Comment (huichen123) [reply build/ProjectReunion-CI.yml:60]: why do you put condition inside parameters ? :)

Comment (getrou) [reply build/ProjectReunion-CI.yml:60]: Since the dependsOn was parameterized, I figured the condition should be as well.

Comment (huichen123) [reply build/ProjectReunion-CI.yml:60]: oh, you have a parameter whose name collides with a keyword. Not sure whether it's a good practice. Try not to pass condition in, but check succeeded() inside CreateNugetPackage directly and see if it makes any difference. In the end, succeeded() check is default and unnecessary.

Comment (getrou) [reply build/ProjectReunion-CI.yml:60]: Same result unfortunately.
https://dev.azure.com/ms/ProjectReunion/_build/results?buildId=114749&view=results

Comment (huichen123) [reply build/ProjectReunion-CI.yml:60]: no clue. I'd say just remove condition as succeeded should be the default behavior.

Comment (getrou) [reply build/ProjectReunion-CI.yml:60]: I believe that will also result in the job being skipped.  I'm pretty confused about why that is, since the previous job that CreateNugetPacakge depends clearly succeeded.

I'm running it without any condition here:
https://dev.azure.com/ms/ProjectReunion/_build/results?buildId=114751&view=results

Comment (getrou) [reply build/ProjectReunion-CI.yml]: Solution was to make the condition specifically look at the PublishMRT job.  I'm not sure why this is, but it works.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 4 above)
Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: I handled this in email, there is a doc (which will be live on github so everyone can see) that explains the packaging plan.  Basically there will be "thin" nuget packages from the various repos that aggregate into Project Reunion and Project Reunion's "metapackage" will reference all of these.  So this repo outputs one "thin" nuget package with metadata and VS targets files, etc. and one metapackage that references everything in Project Reunion.  The actual binaries are aggregated into one framework package that gets used at runtime.

Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: "thin" == metadata plus binaries (dll's), "fat" == contains dll, Project Reunion contributors will publish thin nuget packages to public nuget.org.  The binaries live in the Project Reunion framework package.

Comment (getrou) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: Awesome, thanks Eho.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 6 above)
Comment (EHO-makai) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: this was already working, can write:  https://aka.ms/ProjectReunion :-)

File: build/ProjectReunion-CI.yml (hunk 7 above)
Comment (EHO-makai) [reply build/ProjectReunion-CI.yml]: Is this task going away soon?  I'm wondering if this going to come back ... (no action requested from this comment).

File: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml (hunk 11 above)
Comment (EHO-makai) [reply build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:129]: To be explicit, contact me and let's run the release pipeline together (I have to trigger it, I have the signing authorization).

**Review by EHO-makai — APPROVED** [review]
Consider Hui's sign-off mandatory, but here is mine.

File: dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec (hunk 6 above)
Comment (getrou) [reply dev/MRTCore/mrt/NuGet/Microsoft.ProjectReunion.nuspec]: Yep, done!  This file got moved to build/NuSpecs/Microsoft.ProjectReunion.nuspec, where it points to that link now.

File: build/ProjectReunion-CI.yml (hunk 7 above)
Comment (getrou) [reply build/ProjectReunion-CI.yml]: Is this job going away soon?  I don't think so.  I think even once we start onboarding more sub-modules, this job will stay mostly as-is.

**Comment by huichen123** [comment]
>         outPathRoot: '${{ parameters.nupkgdir }}\signed'

please address this. or you will 2 packages: one signed, one unsigned.

---
In reply to: [705698105](https://github.com/microsoft/ProjectReunion/pull/230#issuecomment-705698105) [](ancestors = 705698105)

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:116 in aa04ba8. [](commit_id = aa04ba83695ef955045b453845fe7eaf744c5c8f, deletion_comment = False)

**Comment by huichen123** [comment]
>       command: 'push'

this only done when we signed the code

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:135 in 28ae59e. [](commit_id = 28ae59e64ef23f54f0a0659a441700bcbccaa7c7, deletion_comment = False)

File: build/NuSpecs/Microsoft.ProjectReunion.nuspec [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+         <license type="file">LICENSE</license>
+         <requireLicenseAcceptance>true</requireLicenseAcceptance>
+         <projectUrl>https://aka.ms/projectreunion</projectUrl>
+         <copyright>© Microsoft Corporation. All rights reserved.</copyright>
+         <tags>Microsoft Resources Native</tags>
```
Comment (huichen123) [inline build/NuSpecs/Microsoft.ProjectReunion.nuspec:13]: >Microsoft Resources Native [](start = 15, length = 26)

add Reunion as tag?

File: build/ProjectReunion-CI.yml [hunk 14]
```diff
@@ -50,8 +55,9 @@ jobs:
 - template: AzurePipelinesTemplates\ProjectReunion-CreateNugetPackage-Job.yml
   parameters:
     jobName: CreateNugetPackage
-    dependsOn: Build
+    dependsOn: PublishMRT
```
Comment (huichen123) [inline build/ProjectReunion-CI.yml:58]: >dependsOn [](start = 4, length = 9)

the dependency in release pipeline also needs change, but that can be in another PR.

**Comment by huichen123** [comment]
in the latest build, there are still binlogs in the package.

**Comment by huichen123** [comment]
also reunion.dll is missing.

File: build/NuSpecs/Microsoft.ProjectReunion.nuspec (hunk 13 above)
Comment (getrou) [reply build/NuSpecs/Microsoft.ProjectReunion.nuspec:13]: I'll add the tags from the old nuspec, "Windows Reunion ProjectReunion"

**Comment by huichen123** [comment]
>       publishVstsFeed: $(NuGetFeed)

where is this defined?

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:141 in 97a2e8c. [](commit_id = 97a2e8c818a7bb5e5cb0226cc002deba12184cb2, deletion_comment = False)

**Comment by getrou** [comment]
> also reunion.dll is missing.

I think we want reunion.dll removed.  It's part of our sample app code, and we don't want it in the nuget package.  The goal is that the only thing in the package is the MRTCore stuff for now.

**Comment by huichen123** [comment]
>       command: 'push'

please address this, or the pipeline will fail.

---
In reply to: [706362132](https://github.com/microsoft/ProjectReunion/pull/230#issuecomment-706362132) [](ancestors = 706362132)

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:135 in 28ae59e. [](commit_id = 28ae59e64ef23f54f0a0659a441700bcbccaa7c7, deletion_comment = False)

**Comment by huichen123** [comment]
>       packagesToPush: '$(Build.ArtifactStagingDirectory)/*.nupkg;!$(Build.ArtifactStagingDirectory)/*.symbols.nupkg'

the pattern is incorrect, thus the correct pipeline skips this

---
Refers to: build/AzurePipelinesTemplates/ProjectReunion-CreateNugetPackage-Job.yml:136 in 60cff54. [](commit_id = 60cff547c08bafddb348b47df2a9d0c37a47bde4, deletion_comment = False)
